### PR TITLE
Reuse headers from THNN in THNN.lua

### PR DESCRIPTION
**Pull request by fmassa**
Instead of having to declare the function headers both in `lib/THNN/generic/THNN.h` and `THNN.lua`, copy the headers from `lib/THNN/generic/THNN.h` into a lua file which returns a string with the content of `lib/THNN/generic/THNN.h`, and use it in `THNN.lua`. 
In the process, strips all lines starting with a `#`.

Also, add missing function declarations to `lib/THNN/generic/THNN.h` (which were present though in `THNN.lua`

Will do the same thing for 'THCUNN` when we agree on the format.

### CODE REVIEW

**Comment by soumith** [comment]
looks good to me!

**Comment by soumith** [comment]
but will wait a couple more days.

**Comment by soumith** [comment]
let's merge this, prep one for THCUNN as well. you need to rebase.

**Comment by fmassa** [comment]
Rebased against master. Will send a PR for THCUNN tomorrow.

**Comment by Atcold** [comment]
LOL, @fmassa, I read "Rebased again, master." Hahaha :laughing:

**Comment by fmassa** [comment]
Thanks @soumith 
I think one of the last bits missing in `THNN` is to move [this part of CMakeLists](https://github.com/torch/nn/blob/master/CMakeLists.txt#L6-L48) to [here](https://github.com/torch/nn/blob/master/lib/THNN/CMakeLists.txt#L11). What do you think ?

**Comment by soumith** [comment]
yep, that makes sense to do!

**Comment by fmassa** [comment]
Ok. Will send a PR

File: THNN.lua [hunk 1]
```diff
@@ -2,994 +2,12 @@ local ffi = require 'ffi'
...
-
-TH_API void THNN_(SpatialReplicationPadding_updateGradInput)(THNNState *state,
-                                                            THTensor *input,
-                                                            THTensor *gradOutput,
-                                                            THTensor *gradInput,
-                                                            int pad_l, int pad_r,
-                                                            int pad_t, int pad_b);
-
-]]
+local generic_THNN_h = require 'nn.THNN_h'
```
Comment (Atcold) [inline THNN.lua]: I'm lost. Where is `THNN_h.lua`?

Comment (soumith) [inline THNN.lua]: can you put some effort and find it. even doing a simple text search on this page will find it for you

Comment (Atcold) [inline THNN.lua]: :unamused: ...
So far I found `lib/THNN/generic/THNN.h` and `THNN.lua`.
There is not any `THNN_h.lua` around.

Comment (fmassa) [inline THNN.lua]: @Atcold It's defined in the [CMakeLists here](https://github.com/torch/nn/pull/674/files#diff-af3b638bc2a3e6c650974192a53c7291R53). It's basically copying `THNN.h` to `THNN_h.lua`

Comment (Atcold) [inline THNN.lua]: Thank you, @fmassa. I was browsing the file tree, and I didn't think it was made on the fly.
I saw also that there are two `THNN.h`... I'm having a look at that now.

**Pull request merged**